hazelcast/src/main/java/com/hazelcast/config/IndexConfig.java

```diff
@@ -0,0 +1,247 @@
+            }
+        }
+    }
+
+    public IndexConfig(IndexConfig other) {
+        this.name = other.name;
+        this.type = other.type;
+
+        for (String attribute : other.getAttributes()) {
+            addAttributeInternal(attribute);
+        }
+    }
+
+    /**
+     * Gets name of the index.
+     *
+     * @return Name of the index or {@code null} if index name should be generated automatically.
+     */
+    public String getName() {
+        return name;
+    }
+
+    /**
+     * Sets name of the index.
+     *
+     * @param name Name of the index or {@code null} if index name should be generated automatically.
+     * @return This instance for chaining.
+     */
+    public IndexConfig setName(String name) {
+        this.name = name;
+
+        return this;
+    }
+
+    /**
+     * Gets type of the index.
+     * <p>
+     * Defaults to {@link IndexType#SORTED}.
+     *
+     * @return Type of the index.
+     */
+    public IndexType getType() {
+        return type;
+    }
+
+    /**
+     * Sets type of the index.
+     * <p>
+     * Defaults to {@link IndexType#SORTED}.
+     *
+     * @param type Type of the index.
+     * @return This instance for chaining.
+     */
+    public IndexConfig setType(IndexType type) {
+        this.type = checkNotNull(type, "Index type cannot be null.");
+
+        return this;
+    }
+
+    /**
+     * Gets index attributes.
+     *
+     * @return Index attributes.
+     */
+    public List<String> getAttributes() {
+        if (attributes == null) {
+            attributes = new ArrayList<>();
+        }
+
+        return attributes;
+    }
+
+    /**
+     * Adds an index attribute with the given.
+     *
+     * @param attribute Attribute name.
+     * @return This instance for chaining.
+     */
+    public IndexConfig addAttribute(String attribute) {
+        addAttributeInternal(attribute);
+
+        return this;
+    }
+
+    public void addAttributeInternal(String attribute) {
+        IndexUtils.validateAttribute(attribute);
+
+        if (attributes == null) {
+            attributes = new ArrayList<>();
+        }
+
+        attributes.add(attribute);
+    }
+
+    /**
+     * Sets index attributes.
+     *
+     * @param attributes Index attributes.
+     * @return This instance for chaining.
+     */
+    public IndexConfig setAttributes(List<String> attributes) {
+        checkNotNull(attributes, "Index attributes cannot be null.");
+
+        this.attributes = new ArrayList<>(attributes.size());
+
+        for (String attribute : attributes) {
+            addAttribute(attribute);
+        }
+
+        return this;
+    }
+
+    @Override
+    public int getFactoryId() {
+        return ConfigDataSerializerHook.F_ID;
+    }
+
+    @Override
+    public int getClassId() {
+        return ConfigDataSerializerHook.INDEX_CONFIG;
+    }
+
+    @Override
+    public void writeData(ObjectDataOutput out) throws IOException {
+        out.writeUTF(name);
```

Comment:
Looks like index type is not persisted and not taken into account in `hashCode`/`equals`, so basically all deserialized index configs are of `SORTED` type.

Reply:
Wow, that is huge. Fixed here: https://github.com/hazelcast/hazelcast/pull/15537/commits/0232915ee3bc8e6ce83d1f25e170a2687a98f0be